test/FunctionalTests/Microsoft.OData.Edm.Tests/Validation/ValidationRulesTests.cs

```diff
@@ -795,6 +795,110 @@ public void SameFunctionsShouldInWithOneInReferencedModelShouldRaiseError()
+
+            ValidateError(
+                ValidationRules.RecursiveComplexTypedPropertyMustBeOptional,
+                nestedProperty,
+                EdmErrorCode.RecursiveComplexTypedPropertyMustBeOptional,
+                Strings.EdmModel_Validator_Semantic_RecursiveComplexTypedPropertyMustBeOptional("nested"));
+        }
+
+        [Fact]
+        public void ComplexTypedPropertyWithSameBaseTypeAsDeclaringTypeShouldError()
+        {
+            EdmComplexType baseType = new EdmComplexType("ns", "myType");
+            EdmComplexType derivedType = new EdmComplexType("ns", "derivedType", baseType);
+            IEdmTypeReference derivedTypeReference = derivedType.GetTypeReference(false);
+            IEdmStructuralProperty nestedBaseProperty = baseType.AddStructuralProperty("nested", derivedTypeReference);
+
+            ValidateError(
+                ValidationRules.RecursiveComplexTypedPropertyMustBeOptional,
+                nestedBaseProperty,
+                EdmErrorCode.RecursiveComplexTypedPropertyMustBeOptional,
+                Strings.EdmModel_Validator_Semantic_RecursiveComplexTypedPropertyMustBeOptional("nested"));
+        }
+
+        [Fact]
+        public void ComplexTypedPropertyWithSameTypeAsDeclaringBaseTypeShouldNotError()
+        {
+            EdmComplexType baseType = new EdmComplexType("ns", "myType");
+            EdmComplexType derivedType = new EdmComplexType("ns", "derivedType", baseType);
+            IEdmTypeReference baseTypeReference = baseType.GetTypeReference(false);
+            IEdmStructuralProperty nestedDerivedProp = derivedType.AddStructuralProperty("nested", baseTypeReference);
+
+            var model = new EdmModel();
+            model.AddElement(baseType);
+            model.AddElement(derivedType);
+
+            ValidateNoError(
+                ValidationRules.RecursiveComplexTypedPropertyMustBeOptional,
+                model,
+                nestedDerivedProp);
+        }
+
+        [Fact]
+        public void ComplexTypedPropertyWithSameInheritedBaseTypeAsDeclaringTypeShouldError()
+        {
+            EdmComplexType a = new EdmComplexType("ns", "a");
+            EdmComplexType b = new EdmComplexType("ns", "b", a);
+            EdmComplexType d = new EdmComplexType("ns", "d", b);
+            IEdmTypeReference dRef = d.GetTypeReference(false);
+            IEdmStructuralProperty aProp = a.AddStructuralProperty("aProp", dRef);
+
+            ValidateError(
+                ValidationRules.RecursiveComplexTypedPropertyMustBeOptional,
+                aProp,
+                EdmErrorCode.RecursiveComplexTypedPropertyMustBeOptional,
+                Strings.EdmModel_Validator_Semantic_RecursiveComplexTypedPropertyMustBeOptional("aProp"));
+        }
+
+        [Fact]
+        public void ComplexTypedPropertyWithSameSecondInheritedBaseTypeAsDeclaringTypeShouldError()
+        {
+            EdmComplexType a = new EdmComplexType("ns", "a");

```

Comment:
It is probably ok using "a", "b", "c". 
but, if we can use more meaningful naming, it could be better.

Reply:
✅